### PR TITLE
Starting Urbit in daemon mode backgrounds the process once the ship has booted.

### PR DESCRIPTION
**Pull request by benjamin-tlon**

### CODE REVIEW

File: vere/main.c [hunk 1]
```diff
@@ -546,6 +553,66 @@ main(c3_i   argc,
...
+      return 1; // TODO
+    }
+
+    fprintf(stderr, "forking\n");
+    fflush(stderr);
+    pid_t pid = fork();
+
+    if (pid) {
+      close(pipefd[1]);
+      close(0);
```
Comment (belisarius222) [inline vere/main.c]: I think we should keep the notation consistent and use `stdin` and `stdout` instead of `0` and `1`.

File: vere/main.c [hunk 2]
```diff
@@ -546,6 +553,66 @@ main(c3_i   argc,
...
+      close(1);
+
+      fprintf(stderr, "parent: waiting\n");
+      fflush(stderr);
+
+      if (read(pipefd[0], buf, 1)) {
+        fprintf(stderr, "parent: urbit booted, exiting\n");
```
Comment (belisarius222) [inline vere/main.c]: This printf is pretty misleading. I'd change it to something more like `"child process started, exiting\r\n"`. Also, I think in general we want to do `\r\n` at the end of `fprintf()` calls, or else they'll show up at the end of the current line.

**Comment by benjamin-tlon** [comment]
@belisarius222 We fork before the terminal gets setup for specifically this reason. The child process will have full access to the file descriptors for the terminal in exactly the same way as it usually does.

And actually, that doesn't matter here, since this only happens in daemon mode, where the terminal stuff is disabled anyways.

**Comment by benjamin-tlon** [comment]
@belisarius222 Tested flows:

- Everything works normally without the `-d` flag.
- The parent process exits with the correct error code if the child process dies during boot.
- The parent process exits on boot if it's successful, and running an urb.py command immediately afterwords works.

**Comment by benjamin-tlon** [comment]
This is ready to merge. Would like feedback on the documentation, and on the location of the `onboot` call.

File: include/vere/vere.h [hunk 3]
```diff
@@ -600,6 +600,7 @@
         c3_i       xit_i;                   //  exit code for shutdown
         void*      tls_u;                   //  server SSL_CTX*
         u3_trac    tra_u;                   //  tracing information
+        void     (*onboot)();               //  call when chis is up
```
Comment (belisarius222) [inline include/vere/vere.h]: I think we should keep the naming scheme for now, in which case this is `bot_f` or something else with three letters and then `_f` for function pointer.

File: vere/main.c [hunk 4]
```diff
@@ -535,6 +543,40 @@ _stop_exit(c3_i int_i)
   u3_pier_exit();
 }
 
+/*
+  This is set the the write-end of a pipe when Urbit is started in daemon
```
Comment (belisarius222) [inline vere/main.c]: grammar: `This is set the write-end of a pipe`

File: vere/reck.c [hunk 5]
```diff
@@ -161,6 +161,14 @@ _reck_kick_http(u3_pier* pir_u,
     {
       u3_http_ef_form(u3k(p_fav));
 
+      // The control server has now started.
+      //
+      // If we're in daemon mode, we need to inform that parent process
```
Comment (belisarius222) [inline vere/reck.c]: that parent process -> the parent process

File: vere/main.c [hunk 6]
```diff
@@ -546,6 +588,59 @@ main(c3_i   argc,
...
+    In both processes, we are good fork() citizens, and close all unused
+    file descriptors. Closing `pipefd[1]` in the parent process is
+    especially important, since the pipe needs to be closed if the child
+    process dies. When the pipe is closed, the read fails, and that's
+    how we know that something went wrong.
+
+    There are some edge cases around `WEXITSTATUS` that are not handled
+    here, but I don't think it matters.
+  */
+  if ( c3y == u3_Host.ops_u.dem ) {
```
Comment (joemfb) [inline vere/main.c]: I think it might be nicer if the contents of this block were in a separate function.

**Comment by belisarius222** [comment]
I agree with Joe's comments and I had my own style nit or two, but the functionality looks good.

**Pull request merged**